### PR TITLE
Add documentation for PageBucketReceiver and PageResultListener

### PR DESCRIPTION
**Pull request by mxm**

### CODE REVIEW

File: sql/src/main/java/io/crate/jobs/PageBucketReceiver.java [hunk 1]
```diff
@@ -56,6 +56,14 @@
  */
 public interface PageBucketReceiver {
 
+    /**
+     * Receives a bucket from an upstream which holds result data. This method may
+     * be called multiple times with the same bucketIdx if there are multiple pages.
```
Comment (seut) [inline sql/src/main/java/io/crate/jobs/PageBucketReceiver.java]: think this could be understand wrongly as calling this with the same bucketIdx on the same page will fail.

File: sql/src/main/java/io/crate/operation/PageResultListener.java [hunk 2]
```diff
@@ -1,7 +1,38 @@
...
+/**
+ * An interface to signal that more paging data is available.
+ * The PageResultListener has to send a request to ask for more data from the source.
+ */
 public interface PageResultListener {
 
+    /**
+     * Indicates that more data can be received. The implementation has to
+     * trigger an action to request more data.
+     * See {@link io.crate.executor.transport.distributed.TransportDistributedResultAction.SendResponsePageResultListener}
```
Comment (seut) [inline sql/src/main/java/io/crate/operation/PageResultListener.java]: not sure if it's a good idea to document implementation details into an interface

File: sql/src/main/java/io/crate/operation/PageResultListener.java [hunk 3]
```diff
@@ -1,7 +1,38 @@
...
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
 
+package io.crate.operation;
 
+/**
+ * An interface to signal that more paging data is available.
```
[the remark this answers is not in the document]
Comment (mxm) [reply sql/src/main/java/io/crate/operation/PageResultListener.java]: Thanks for your suggestion, I phrased it slightly different in the method doc and will also adjust here.

File: sql/src/main/java/io/crate/operation/PageResultListener.java (hunk 2 above)
Comment (mxm) [reply sql/src/main/java/io/crate/operation/PageResultListener.java]: As example I think it is ok. What else than triggering something external can this method do, since it's returning void?

File: sql/src/main/java/io/crate/jobs/PageBucketReceiver.java (hunk 1 above)
Comment (mxm) [reply sql/src/main/java/io/crate/jobs/PageBucketReceiver.java]: As far as I understand, for every page it can be called once but wouldn't that be implementation detail also? :)

Comment (seut) [reply sql/src/main/java/io/crate/jobs/PageBucketReceiver.java]: yes so i'd rephrase it or leave it out

File: sql/src/main/java/io/crate/operation/PageResultListener.java (hunk 2 above)
Comment (seut) [reply sql/src/main/java/io/crate/operation/PageResultListener.java]: throwing an exception? its all up to the implementation

Comment (mfussenegger) [reply sql/src/main/java/io/crate/operation/PageResultListener.java]: How about referencing back to https://github.com/crate/crate/pull/6681/files#diff-173f7d1183bba8fca495fad1939d503fR53 if `needMore(true)` is called, the PageBucketReceiver expects another `setBucket` calls. If `needMore=false` nothing has to be done by the listener implementation and it can free up any resources it had acquired as it won't be needed anymore

File: sql/src/main/java/io/crate/jobs/PageBucketReceiver.java [hunk 4]
```diff
@@ -56,6 +56,14 @@
...
+     * @param bucketIdx A bucket id which uniquely identifies all buckets of this page.
+     * @param rows The bucket which holds result rows.
+     * @param isLast Indicates whether this is the last bucket with this id.
+     * @param pageResultListener The ResultListener which is informed if more data is available.
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/jobs/PageBucketReceiver.java]: this is missleading, it's not if more data is available, it's that more data is required. (As stated in https://github.com/crate/crate/pull/6681/files#diff-173f7d1183bba8fca495fad1939d503fR53)

Comment (mxm) [reply sql/src/main/java/io/crate/jobs/PageBucketReceiver.java]: It's both at the same time because the context knows that it hasn't received the `isLast` flag yet. Shouldn't go here in the interface though. I'll rephrase to make it clear.

File: sql/src/main/java/io/crate/jobs/PageBucketReceiver.java [hunk 5]
```diff
@@ -56,6 +56,14 @@
  */
 public interface PageBucketReceiver {
 
+    /**
+     * Receives a bucket from an upstream which holds result data. This method may
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/jobs/PageBucketReceiver.java]: I think the contract is as follows:

This is called *once* per bucketIdx initially; then exactly once for each `bucketIdx/pageResultListener.needMore=true`

(The important difference to what is currently documented in this PR is that it is *not* expected to get another call if `needMore=false` for that bucketIdx)

File: sql/src/main/java/io/crate/operation/PageResultListener.java [hunk 6]
```diff
@@ -1,7 +1,37 @@
...
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
 
+package io.crate.operation;
 
+/**
+ * An interface to signal more paging data is needed.
+ * The PageResultListener has to take action (e.g. send a request to ask for more data from the source).
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/PageResultListener.java]: What the PageResultListener does is not relevant, what's important is that it results in another `setBucket` or `setFailure` call (if `needMore=true`)

Comment (mxm) [reply sql/src/main/java/io/crate/operation/PageResultListener.java]: Thanks, I've rephrased this to reflect your suggestion.

File: sql/src/main/java/io/crate/jobs/PageBucketReceiver.java (hunk 5 above)
Comment (mxm) [reply sql/src/main/java/io/crate/jobs/PageBucketReceiver.java]: Thanks, I've rephrased this to reflect your suggestion.
